pkg/rsnotify/listener/ipreporter.go

```diff
@@ -0,0 +1,62 @@
+package listener
+
+// Copyright (C) 2022 by RStudio, PBC.
+
+import "net"
+
+type IPReporter interface {
+	IP() string
+}
+
+type IPCache struct {
+	reporter    IPReporter
+	cachedValue string
+}
+
+// Note that this also implements the `IPReporter` interface, so there's really no need
+// to have a separate interface for the cache.
+func (c *IPCache) IP() string {
+	gotIP := c.reporter.IP()
+	if gotIP != c.cachedValue {
+		// enumerate IPs and determine if cached value is still value
+		if !c.stillValidIP(c.cachedValue) {
+			c.cachedValue = gotIP
+		}
+	}
+
+	return c.cachedValue
+}
+
+func (p *IPCache) stillValidIP(IP string) bool {
+	var ip net.IP
+	if ip = net.ParseIP(IP); ip == nil {
+		return false
+	}
+
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return false
+	}
+
+	for _, inter := range interfaces {
+		addrs, err := inter.Addrs()
+		if err != nil {
+			return false
+		}
+
+		for _, addr := range addrs {
+			switch v := addr.(type) {
+			case *net.IPNet:
+				if v.Contains(ip) {
+					return true
+				}
+			case *net.IPAddr:
```

Comment:
Can you add a comment that mentions that `net.IPAddr` is sometimes returned by Windows? I don't see any case on BSD, Linux, or Darwin that returns one of these.

Reply:
According to [interface_linux.go](https://cs.opensource.google/go/go/+/refs/tags/go1.17.8:src/net/interface_linux.go;l=239) it's possible to see IPAddr on Linux.